File: src/condash/paths.py
"""Path validation for the conception directory tree.

Every HTTP-facing route that accepts a user-supplied path uses one of the
validators here to resolve it safely under ``ctx.base_dir``. The shared
``_safe_resolve`` helper consolidates the "reject traversal + regex-gate +
resolve + relative_to + existence check" dance that every validator used
to duplicate inline.
"""

from __future__ import annotations

import re
from collections.abc import Sequence
from pathlib import Path

from .context import RenderCtx

_VALID_ITEM_PREFIX = r"^projects/\d{4}-\d{2}/\d{4}-\d{2}-\d{2}-[\w.-]+/"

_VALID_PATH_RE = re.compile(_VALID_ITEM_PREFIX + r"README\.md$")

# Any file at any depth under an item directory — cards now surface the
# full item tree, not just ``notes/``, so the regexes that gate file
# serving must follow. Traversal is still blocked by ``_safe_resolve``
# (the ".." reject + relative_to check), so relaxing the regex to an
# arbitrary subpath stays safe.
_VALID_DOWNLOAD_RE = re.compile(_VALID_ITEM_PREFIX + r"[\w./-]+\.pdf$")

_VALID_NOTE_RE = re.compile(_VALID_ITEM_PREFIX + r"[\w./-]+\.md$")

# Knowledge pages live outside the date-prefixed item structure. Match
# `knowledge/<file>.md` at the root (conventions.md) and any depth of
# subdir under it (`topics/index.md`, `topics/ops/dev-ports.md`, …).
# `_safe_resolve` still rejects traversal (`..`, `\x00`) and enforces
# `relative_to(base)`, so allowing arbitrary depth here stays safe.
_VALID_KNOWLEDGE_NOTE_RE = re.compile(r"^knowledge/(?:[\w.-]+/)*[\w.-]+\.md$")

_VALID_ASSET_RE = re.compile(
    _VALID_ITEM_PREFIX + r"[\w./-]+\.(?:png|jpg|jpeg|gif|svg|webp)$",
    re.IGNORECASE,
)

# Any file at any depth inside an item directory.
_VALID_ITEM_FILE_RE = re.compile(_VALID_ITEM_PREFIX + r"[\w./-]+$")

# Restricted to files directly under ``<item>/notes/`` — used by rename
# (only notes are user-renamable; loose item files and READMEs are not).
_VALID_ITEM_NOTES_FILE_RE = re.compile(_VALID_ITEM_PREFIX + r"notes/[\w./-]+$")

_VALID_NOTE_FILENAME_RE = re.compile(r"^[\w.-]+\.[A-Za-z0-9]+$")

_ASSET_CONTENT_TYPES = {
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".gif": "image/gif",
    ".svg": "image/svg+xml",
    ".webp": "image/webp",
}


def _safe_resolve(
    ctx: RenderCtx,
    rel_path: str,
    regexes: Sequence[re.Pattern] = (),
    *,
    require_file: bool = True,
    strict: bool = False,
) -> Path | None:
    """Resolve ``rel_path`` under ``ctx.base_dir``, rejecting anything outside."""
    if not rel_path or "\x00" in rel_path or ".." in rel_path:
        return None
    if regexes and not any(p.match(rel_path) for p in regexes):
        return None
    base = ctx.base_dir
    try:
        if strict:
            full = (base / rel_path).resolve(strict=True)
        else:
            full = (base / rel_path).resolve()
        full.relative_to(base.resolve())
    except (OSError, ValueError):
        return None
    if require_file:
        return full if full.is_file() else None
    return full if full.exists() else None


def _validate_path(ctx: RenderCtx, rel_path: str) -> Path | None:
    """Validate an item-README path."""
    return _safe_resolve(ctx, rel_path, (_VALID_PATH_RE,), require_file=False)


def validate_note_path(ctx: RenderCtx, rel_path: str) -> Path | None:
    """Public: validate a note/README/knowledge/notes-file path."""
    return _safe_resolve(
        ctx,
        rel_path,
        (_VALID_NOTE_RE, _VALID_KNOWLEDGE_NOTE_RE, _VALID_ITEM_FILE_RE),
    )


def _validate_doc_path(ctx: RenderCtx, rel_path: str) -> Path | None:
    """Resolve a note-body link target against the conception tree."""
    return _safe_resolve(ctx, rel_path, strict=True)


_VALID_ITEM_DIR_RE = re.compile(r"^projects/\d{4}-\d{2}/\d{4}-\d{2}-\d{2}-[\w.-]+/?$")


def _validate_item_dir(ctx: RenderCtx, rel_path: str) -> Path | None:
    """Resolve a project-item folder path against the conception tree."""
    full = _safe_resolve(ctx, rel_path, (_VALID_ITEM_DIR_RE,), require_file=False)
    if full is None or not full.is_dir():
        return None
    return full


def validate_download_path(ctx: RenderCtx, rel_path: str) -> Path | None:
    return _safe_resolve(ctx, rel_path, (_VALID_DOWNLOAD_RE,))


def validate_asset_path(ctx: RenderCtx, rel_path: str) -> tuple[Path, str] | None:
    full = _safe_resolve(ctx, rel_path, (_VALID_ASSET_RE,))
    if full is None:
        return None
    ctype = _ASSET_CONTENT_TYPES.get(full.suffix.lower(), "application/octet-stream")
    return full, ctype


def validate_file_path(ctx: RenderCtx, rel_path: str) -> tuple[Path, str] | None:
    """Validate a raw-byte serve request for the /file endpoint."""
    full = validate_note_path(ctx, rel_path)
    if full is None:
        return None
    return full, _guess_content_type(full)


def _guess_content_type(path: Path) -> str:
    import mimetypes

    ext = path.suffix.lower()
    if ext in _ASSET_CONTENT_TYPES:
        return _ASSET_CONTENT_TYPES[ext]
    if ext == ".pdf":
        return "application/pdf"
    if ext == ".md":
        return "text/markdown; charset=utf-8"
    guess, _ = mimetypes.guess_type(path.name)
    return guess or "application/octet-stream"


def _validate_open_path(ctx: RenderCtx, path_str: str) -> Path | None:
    """Validate an absolute filesystem path against the workspace sandbox."""
    if not path_str or "\x00" in path_str:
        return None
    try:
        p = Path(path_str).resolve(strict=True)
    except (OSError, RuntimeError):
        return None
    if not p.is_dir():
        return None
    roots: list[Path] = []
    if ctx.workspace is not None:
        roots.append(ctx.workspace.resolve())
    if ctx.worktrees is not None:
        roots.append(ctx.worktrees.resolve())
    for root in roots:
        try:
            p.relative_to(root)
            return p
        except ValueError:
            continue
    return None

File: src-tauri/src/paths.rs
//! Path validation for user-supplied paths used by the write-side
//! routes: a `safe_resolve` helper plus regex-gated validators for
//! READMEs, notes, item dirs, and filenames.
//!
//! Every validator follows the same four-step shape:
//!
//! 1. Reject empty / NUL / literal `..` outright.
//! 2. Gate with one or more regexes.
//! 3. Resolve under `base_dir`, refuse anything whose canonical path
//!    escapes it.
//! 4. Optionally require the resolved entry be a file on disk.

use std::path::{Path, PathBuf};
use std::sync::LazyLock;

use regex::Regex;

/// Common item-path prefix — `projects/YYYY-MM/YYYY-MM-DD-<slug>/`.
const VALID_ITEM_PREFIX: &str = r"^projects/\d{4}-\d{2}/\d{4}-\d{2}-\d{2}-[\w.-]+/";

/// `projects/YYYY-MM/YYYY-MM-DD-<slug>/README.md`.
static VALID_README_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!("{VALID_ITEM_PREFIX}README\\.md$")).expect("VALID_README_RE compiles")
});

/// `<item>/<anything>.md`.
static VALID_NOTE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(r"{VALID_ITEM_PREFIX}[\w./-]+\.md$")).expect("VALID_NOTE_RE compiles")
});

/// `knowledge/<file>.md`. Matches `knowledge/` at any depth.
static VALID_KNOWLEDGE_NOTE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^knowledge/(?:[\w.-]+/)*[\w.-]+\.md$").expect("VALID_KNOWLEDGE_NOTE_RE compiles")
});

/// Any file at any depth inside an item directory.
static VALID_ITEM_FILE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(r"{VALID_ITEM_PREFIX}[\w./-]+$")).expect("VALID_ITEM_FILE_RE compiles")
});

/// Restricted to files directly under `<item>/notes/` — used by rename
/// (only notes are user-renamable).
pub(crate) static VALID_ITEM_NOTES_FILE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(r"{VALID_ITEM_PREFIX}notes/[\w./-]+$"))
        .expect("VALID_ITEM_NOTES_FILE_RE compiles")
});

/// Item directory itself — `projects/YYYY-MM/YYYY-MM-DD-<slug>/`, with
/// or without the trailing slash.
static VALID_ITEM_DIR_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^projects/\d{4}-\d{2}/\d{4}-\d{2}-\d{2}-[\w.-]+/?$")
        .expect("VALID_ITEM_DIR_RE compiles")
});

/// `<name>.<ext>` — validates the target filename of create_note /
/// rename_note.
pub static VALID_NOTE_FILENAME_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[\w.-]+\.[A-Za-z0-9]+$").expect("VALID_NOTE_FILENAME_RE compiles")
});

/// Bare filename allowed for the rename target's new stem.
pub static VALID_NEW_STEM_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[\w.-]+$").expect("VALID_NEW_STEM_RE compiles"));

/// Subdirectory path (possibly nested). Used by `resolve_under_item`
/// to reject anything that doesn't look like a sequence of `[\w.-]+`
/// segments.
pub static VALID_SUBDIR_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[\w.-]+(/[\w.-]+)*$").expect("VALID_SUBDIR_RE compiles"));

/// Upload filename regex — more permissive than the note regex to
/// accept typical Camera/PDF exports (spaces, parentheses).
pub static VALID_UPLOAD_FILENAME_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[\w. \-()]+\.[A-Za-z0-9]+$").expect("VALID_UPLOAD_FILENAME_RE compiles")
});

/// Low-level: resolve `rel_path` under `base_dir`, rejecting anything
/// outside. `require_file=true` additionally checks that the resolved
/// entry exists as a regular file; `require_file=false` only requires
/// the path to resolve (which on Unix implies existence, since we go
/// through `canonicalize`).
pub fn safe_resolve(
    base: &Path,
    rel_path: &str,
    regexes: &[&Regex],
    require_file: bool,
) -> Option<PathBuf> {
    if rel_path.is_empty() || rel_path.contains('\0') || rel_path.contains("..") {
        return None;
    }
    if !regexes.is_empty() && !regexes.iter().any(|r| r.is_match(rel_path)) {
        return None;
    }
    let candidate = base.join(rel_path);
    let canonical = std::fs::canonicalize(&candidate).ok()?;
    let base_canonical = std::fs::canonicalize(base).ok()?;
    if !canonical.starts_with(&base_canonical) {
        return None;
    }
    if require_file && !canonical.is_file() {
        return None;
    }
    Some(canonical)
}

/// Validate a README path (`projects/YYYY-MM/<slug>/README.md`). The
/// README must resolve to an existing file under `base_dir`. Every
/// step-mutation handler reads the file immediately after, so
/// `require_file=true` is hardcoded.
pub fn validate_readme_path(base_dir: &Path, rel_path: &str) -> Option<PathBuf> {
    safe_resolve(base_dir, rel_path, &[&VALID_README_RE], true)
}

/// Validate a note-like path — item-tree file, `<item>/notes/*.md`, or
/// `knowledge/**/*.md`. The target must resolve to an existing file.
pub fn validate_note_path(base_dir: &Path, rel_path: &str) -> Option<PathBuf> {
    safe_resolve(
        base_dir,
        rel_path,
        &[
            &VALID_NOTE_RE,
            &VALID_KNOWLEDGE_NOTE_RE,
            &VALID_ITEM_FILE_RE,
        ],
        true,
    )
}

/// Validate a project-item folder path (`projects/YYYY-MM/<slug>/`) under
/// `base_dir`. Used by `/open-folder` so the "open folder in file
/// manager" card button resolves against the conception tree (not the
/// workspace/worktrees sandbox).
pub fn validate_item_dir(base_dir: &Path, rel_path: &str) -> Option<PathBuf> {
    let full = safe_resolve(base_dir, rel_path, &[&VALID_ITEM_DIR_RE], false)?;
    if !full.is_dir() {
        return None;
    }
    Some(full)
}

/// Resolve `subdir` (relative to `item_dir`) and verify the result
/// stays inside the item directory. Empty `subdir` resolves to
/// `item_dir` itself. Returns `None` on traversal / regex failure.
pub fn resolve_under_item(item_dir: &Path, subdir: &str) -> Option<PathBuf> {
    let trimmed = subdir.trim().trim_matches('/');
    if trimmed.is_empty() {
        return Some(item_dir.to_path_buf());
    }
    if trimmed.split('/').any(|seg| seg == "..") {
        return None;
    }
    if !VALID_SUBDIR_RE.is_match(trimmed) {
        return None;
    }
    let target = item_dir.join(trimmed);
    // Use `canonicalize` when the target exists, and a literal
    // prefix check when it does not (the target may legitimately not
    // exist yet in the create_note / mkdir flows — the regex + `..`
    // reject above is enough to keep us safely inside).
    if let Ok(canonical_target) = std::fs::canonicalize(&target) {
        let canonical_base = std::fs::canonicalize(item_dir).ok()?;
        if !canonical_target.starts_with(&canonical_base) {
            return None;
        }
        return Some(canonical_target);
    }
    Some(target)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    fn seeded_tree() -> (TempDir, PathBuf) {
        let dir = TempDir::new().unwrap();
        let base = dir.path().to_path_buf();
        let item = base.join("projects/2026-04/2026-04-22-demo");
        fs::create_dir_all(item.join("notes")).unwrap();
        fs::write(item.join("README.md"), "# demo\n").unwrap();
        fs::write(item.join("notes/first.md"), "hello\n").unwrap();
        fs::create_dir_all(base.join("knowledge/topics")).unwrap();
        fs::write(base.join("knowledge/conventions.md"), "c\n").unwrap();
        fs::write(base.join("knowledge/topics/dev.md"), "d\n").unwrap();
        (dir, base)
    }

    #[test]
    fn accepts_well_formed_readme() {
        let (_tmp, base) = seeded_tree();
        assert!(
            validate_readme_path(&base, "projects/2026-04/2026-04-22-demo/README.md").is_some()
        );
    }

    #[test]
    fn rejects_traversal() {
        let (_tmp, base) = seeded_tree();
        assert!(validate_readme_path(&base, "projects/../etc/passwd").is_none());
        assert!(validate_note_path(&base, "projects/../../etc/passwd").is_none());
    }

    #[test]
    fn rejects_wrong_shape() {
        let (_tmp, base) = seeded_tree();
        assert!(validate_readme_path(&base, "README.md").is_none());
        assert!(validate_readme_path(&base, "projects/2026-04/demo/README.md").is_none());
        assert!(validate_readme_path(&base, "notes/some.md").is_none());
    }

    #[test]
    fn rejects_missing_file() {
        let (_tmp, base) = seeded_tree();
        assert!(
            validate_readme_path(&base, "projects/2026-04/2026-04-22-ghost/README.md").is_none()
        );
    }

    #[test]
    fn rejects_nul_and_empty() {
        let (_tmp, base) = seeded_tree();
        assert!(validate_readme_path(&base, "").is_none());
        assert!(validate_readme_path(&base, "a\0b").is_none());
    }

    #[test]
    fn accepts_note_under_item_notes_dir() {
        let (_tmp, base) = seeded_tree();
        assert!(
            validate_note_path(&base, "projects/2026-04/2026-04-22-demo/notes/first.md").is_some()
        );
    }

    #[test]
    fn accepts_knowledge_note() {
        let (_tmp, base) = seeded_tree();
        assert!(validate_note_path(&base, "knowledge/conventions.md").is_some());
        assert!(validate_note_path(&base, "knowledge/topics/dev.md").is_some());
    }

    #[test]
    fn valid_item_notes_file_re_gates_rename_source() {
        assert!(
            VALID_ITEM_NOTES_FILE_RE.is_match("projects/2026-04/2026-04-22-demo/notes/first.md")
        );
        assert!(!VALID_ITEM_NOTES_FILE_RE.is_match("projects/2026-04/2026-04-22-demo/README.md"));
        assert!(!VALID_ITEM_NOTES_FILE_RE.is_match("projects/2026-04/2026-04-22-demo/loose.md"));
    }

    #[test]
    fn resolve_under_item_empty_is_item_dir() {
        let (_tmp, base) = seeded_tree();
        let item = base.join("projects/2026-04/2026-04-22-demo");
        assert_eq!(resolve_under_item(&item, ""), Some(item.clone()));
        assert_eq!(resolve_under_item(&item, "   "), Some(item.clone()));
        assert_eq!(resolve_under_item(&item, "/"), Some(item.clone()));
    }

    #[test]
    fn resolve_under_item_nested() {
        let (_tmp, base) = seeded_tree();
        let item = base.join("projects/2026-04/2026-04-22-demo");
        let got = resolve_under_item(&item, "notes").expect("notes exists");
        assert!(got.ends_with("projects/2026-04/2026-04-22-demo/notes"));
    }

    #[test]
    fn resolve_under_item_rejects_traversal() {
        let (_tmp, base) = seeded_tree();
        let item = base.join("projects/2026-04/2026-04-22-demo");
        assert!(resolve_under_item(&item, "..").is_none());
        assert!(resolve_under_item(&item, "notes/..").is_none());
        assert!(resolve_under_item(&item, "../../etc").is_none());
    }

    #[test]
    fn resolve_under_item_allows_nonexistent_nested() {
        let (_tmp, base) = seeded_tree();
        let item = base.join("projects/2026-04/2026-04-22-demo");
        // The target doesn't exist yet — create_notes_subdir uses this
        // code path. Should still resolve, since the regex + ".." reject
        // keep us safe.
        let got = resolve_under_item(&item, "freshly/nested").expect("ok");
        assert!(got.ends_with("projects/2026-04/2026-04-22-demo/freshly/nested"));
    }

    #[test]
    fn valid_new_stem_accepts_reasonable_names() {
        for good in ["report", "report.draft", "report-2", "under_score"] {
            assert!(VALID_NEW_STEM_RE.is_match(good), "want match: {good}");
        }
        for bad in ["", "has space", "a/b", "../x"] {
            assert!(!VALID_NEW_STEM_RE.is_match(bad), "want reject: {bad}");
        }
    }
}
